### PR TITLE
[Reclaim buffer] Reclaim unused buffer for dynamic buffer model

### PR DESCRIPTION
**Pull request by stephenxs**
**What I did**

Reclaim reserved buffer of unused ports for both dynamic and traditional models.
This is done by 
- Removing lossless priority groups on unused ports.
- Applying zero buffer profiles on the buffer objects of unused ports.
- In the dynamic buffer model, the zero profiles are loaded from a JSON file and applied to `APPL_DB` if there are admin down ports.
  The default buffer configuration will be configured on all ports. Buffer manager will apply zero profiles on admin down ports.
- In the static buffer model, the zero profiles are loaded by the buffer template.

Signed-off-by: Stephen Sun <stephens@nvidia.com>

**Why I did it**

**How I verified it**

Regression test and vs test.

**Details if related**
***Static buffer model***

Remove the lossless buffer priority group if the port is admin-down and the buffer profile aligns with the speed and cable length of the port.

***Dynamic buffer model***

****Handle zero buffer pools and profiles****

1. buffermgrd: add a CLI option to load the JSON file for zero profiles. (done in PR #1996)
2. Load them from JSON file into the internal buffer manager's data structure (done in PR #1996)
3. Apply them to APPL_DB once there is at least one admin-down port
   - Record zero profiles' names in the pool object it references.
     By doing so, the zero profile lists can be constructed according to the normal profile list. There should be one profile for each pool on the ingress/egress side.
   - And then apply the zero profiles to the buffer objects of the port.
   - Unload them from APPL_DB once all ports are admin-up since the zero pools and profiles are no longer referenced.
     Remove buffer pool counter id when the zero pool is removed.
4. Now that it's possible that a pool will be removed from the system, the watermark counter of the pool is removed ahead of the pool itself being removed.

****Handle port admin status change****

1. Currently, there is a logic of removing buffer priority groups of admin down ports. This logic will be reused and extended for all buffer objects, including `BUFFER_QUEUE`, `BUFFER_PORT_INGRESS_PROFILE_LIST`, and `BUFFER_PORT_EGRESS_PROFILE_LIST`.
   - When the port is admin down,
     - The normal profiles are removed from the buffer objects of the port
     - The zero profiles, if provided, are applied to the port
   - When the port is admin up,
     - The zero profiles, if applied, are removed from the port
     - The normal profiles are applied to the port.
2. Ports orchagent exposes the number of queues and priority groups to STATE_DB.
   Buffer manager can take advantage of these values to apply zero profiles on all the priority groups and queues of the admin-down ports.
   In case it is not necessary to apply zero profiles on all priority groups or queues on a certain platform, `ids_to_reclaim` can be customized in the JSON file.
3. Handle all buffer tables, including `BUFFER_PG`, `BUFFER_QUEUE`, `BUFFER_PORT_INGRESS_PROFILE_LIST` and `BUFFER_PORT_EGRESS_PROFILE_LIST`
   - Originally, only the `BUFFER_PG` table was cached in the dynamic buffer manager.
   - Now, all tables are cached in order to apply zero profiles when a port is admin down and apply normal profiles when it's up.
   - The index of such tables can include a single port or a list of ports, like `BUFFER_PG|Ethernet0|3-4` or `BUFFER_PG|Ethernet0,Ethernet4,Ethernet8|3-4`. Originally, there is a logic to handle such indexes for the `BUFFER_PG` table. Now it is reused and extended to handle all the tables.
4. [Mellanox] Plugin to calculate buffer pool size:
   - Originally, buffer for the queue, buffer profile list, etc. were not reclaimed for admin-down ports so they are reserved for all ports.
   - Now, they are reserved for admin-up ports only.

****Accelerate the progress of applying buffer tables to APPL_DB****

This is an optimization on top of reclaiming buffer.

1. Don't apply buffer profiles, buffer objects to `APPL_DB` before buffer pools are applied when the system is starting.
   This is to apply the items in an order from referenced items to referencing items and try to avoid buffer orchagent retrying due to referenced table items.
   However, it is still possible that the referencing items are handled before referenced items. In that case, there should not be any error message.
2. [Mellanox] Plugin to calculate buffer pool size:
   Return the buffer pool sizes value currently in APPL_DB if the pool sizes are not able to be calculated due to lacking some information. This typically happens at the system start.
   This is to accelerate the progress of pushing tables to APPL_DB.

### CODE REVIEW

**Comment by stephenxs** [comment]
Vs test failed by test_PerfAddRemoveRoute 
```
test_PerfAddRemoveRoute failed (1 runs remaining out of 2).
	<class 'AssertionError'>
	Unexpected number of keys: expected=10007, received=10008 (('{"dest":"100.30.143.0/24","switch_id":"oid:0x21000000000000","vr":"oid:0x3000000000022"}', ...))
	[<TracebackEntry /home/vsts/work/1/s/tests/test_route.py:949>, <TracebackEntry /home/vsts/work/1/s/tests/dvslib/dvs_database.py:366>]
test_PerfAddRemoveRoute failed; it passed 0 out of the required 1 times.
	<class 'AssertionError'>
	Unexpected number of keys: expected=20008, received=10008 (('{"dest":"100.30.143.0/24","switch_id":"oid:0x21000000000000","vr":"oid:0x3000000000022"}', ...))
	[<TracebackEntry /home/vsts/work/1/s/tests/test_route.py:949>, <TracebackEntry /home/vsts/work/1/s/tests/dvslib/dvs_database.py:366>]
```

File: cfgmgr/buffermgrdyn.cpp [hunk 1]
```diff
@@ -1874,15 +2506,202 @@ task_process_status BufferMgrDynamic::handleBufferProfileTable(KeyOpFieldsValues
...
+                supportedNotConfiguredItems.erase(idStr);
+            }
+
+            auto const combinedIdsStr = generateIdListFromMap(idsToDelMap, portInfo.maximum_buffer_objects[direction]);
+            // combinedIdsStr should contain only one id string.
+            for (auto const &idStr : combinedIdsStr)
+            {
+                SWSS_LOG_INFO("Removing buffer %s item %s, got combined item %s",
+                              objectName.c_str(), key.c_str(), idStr.c_str());
+                supportedNotConfiguredItems.insert(idStr);
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp:2704]: No need. If zero profiles is not provided, the `supported_but_not_configured_items` will not be generated.
Will add comments to explain it.

**Review by keboliu — DISMISSED** [review]

**Review by neethajohn — APPROVED** [review]

**Comment by stephenxs** [comment]
The retrying will probably still fail today since the required commit ([b3ccef9](https://github.com/Azure/sonic-buildimage/commit/b3ccef9c082ce115efaa6af4c44f11f99a994996)) is still not included in the latest build([53991](https://dev.azure.com/mssonic/build/_build/results?buildId=53991&view=results)) of vs image.
I suppose it will be included in tomorrow’s build so it should pass all checkers tomorrow.

**Review by keboliu — APPROVED** [review]

**Review by neethajohn — APPROVED** [review]

**Pull request merged**